### PR TITLE
AP DDS Tests

### PR DESCRIPTION
**Pull request by SantamRC**
Adding Tests as per #25711

### CODE REVIEW

**Comment by arshPratap** [comment]
Hi @SantamRC  great work ! I have added a few suggestions for this. Please take a look.
Thanks !!

**Comment by Ryanf55** [comment]
Thanks! Looks good, but seems the test is failing in CI. Any ideas?
https://github.com/ArduPilot/ardupilot/actions/runs/7623561288/job/20763840428?pr=26069#step:13:8

**Review by arshPratap — APPROVED** [review]
LGTM ! Good work @SantamRC

**Comment by arshPratap** [comment]
No idea why the CI test is failing

**Comment by SantamRC** [comment]
@arshPratap @Ryanf55 is there a map table to check which number corresponds to which mode? For the latest commit, it would be useful.
Please review.

**Comment by Ryanf55** [comment]
> @arshPratap @Ryanf55 is there a map table to check which number corresponds to which mode? For the latest commit, it would be useful. Please review.

Not that I've found unless we add a dependency on mavproxy: 
https://github.com/ArduPilot/MAVProxy/blob/c4619af111761135408476d6a7e79cea68cf8388/MAVProxy/modules/mavproxy_chat/assistant_setup/copter_flightmodes.txt

FOr now, just define them at the top of the test. I can figure out a better solution long term later.

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_mode_switch.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import rclpy
+import time
+from ardupilot_msgs.srv import ModeSwitch
+
+
+def test_mode_switch():
+    rclpy.init()
+
+    modes = [
```
Comment (arshPratap) [inline Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_mode_switch.py:9]: Hi ! Can we have the `modes` array as an enum.
Thanks

**Comment by arshPratap** [comment]
> > @arshPratap @Ryanf55 is there a map table to check which number corresponds to which mode? For the latest commit, it would be useful. Please review.
> 
> Not that I've found unless we add a dependency on mavproxy: https://github.com/ArduPilot/MAVProxy/blob/c4619af111761135408476d6a7e79cea68cf8388/MAVProxy/modules/mavproxy_chat/assistant_setup/copter_flightmodes.txt
> 
> FOr now, just define them at the top of the test. I can figure out a better solution long term later. They are autopilot specific, and there doesn't seem to be a way to query the mappings at runtime. https://mavlink.io/en/messages/common.html#MAV_CMD_DO_SET_MODE

Hi @SantamRC  , you can follow it based on the links provided [here](https://github.com/ArduPilot/ardupilot/blob/master/Tools/ros2/ardupilot_msgs/srv/ModeSwitch.srv#L4C1-L4C59)

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_mode_switch.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import rclpy
+import time
+from ardupilot_msgs.srv import ModeSwitch
+from enum import Enum
+
+
+class FlightModes(Enum):
```
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_mode_switch.py:7]: Are these copter specific?

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_clock.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            assert isinstance(msg.clock, Time)
+            assert isinstance(msg.clock.sec, int)
+            assert isinstance(msg.clock.nanosec, int)
+
+            print("Assertion passed. Received valid Clock message:")
+            print(f"Seconds: {msg.clock.sec}")
+            print(f"Nanoseconds: {msg.clock.nanosec}")
+
+        except AssertionError as e:
+            print(f"Assertion failed. Invalid Clock message: {e}")
```
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_clock.py:23]: Won't the test now pass even if it fails the assertion?

**Review by Ryanf55 — CHANGES_REQUESTED** [review]

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_mode_switch.py (hunk 2 above)
Comment (SantamRC) [reply Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_mode_switch.py:7]: I have added all the flight modes given here - https://github.com/ArduPilot/MAVProxy/blob/c4619af111761135408476d6a7e79cea68cf8388/MAVProxy/modules/mavproxy_chat/assistant_setup/copter_flightmodes.txt

Should I only include the copter modes?

Comment (Ryanf55) [reply Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_mode_switch.py:7]: Well that list is copter-specific in the title, so perhaps the class  could be called `FlightModesCopter`.  It does not have FBWA, Thermal, Cruise, or any of the other plane-specific modes.

**Pull request closed without merging**